File: Week_09/32/longest_valid_parentheses3.py
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        ans = 0
        left = right = 0

        for i in range(len(s)):
            if s[i] == '(':
                left += 1
            else:
                right += 1
            if left == right:
                ans = max(ans, 2 * left)
            elif left < right:
                left = right = 0
        left = right = 0

        for i in range(len(s) - 1, -1, -1):
            if s[i] == '(':
                left += 1
            else:
                right += 1

            if left == right:
                ans = max(ans, 2 * left)
            elif right < left:
                left = right = 0
        return ans

    def test(self):
        s = "(())"
        ans = self.longestValidParentheses(s)
        target = 4
        assert ans == target, f"target: {target}, ans: {ans}"


if __name__ == "__main__":
    Solution().test()

File: Week_09/32/longest_valid_parentheses3.go
package main

func longestValidParentheses3(s string) int {
	left, right, maxLen := 0, 0, 0
	for i := 0; i < len(s); i++ {
		if s[i] == '(' {
			left++
		} else {
			right++
		}
		if left == right {
			maxLen = max(maxLen, 2*left)
		} else if left < right {
			left = 0
			right = 0
		}
	}
	left, right = 0, 0
	for i := len(s) - 1; i >= 0; i-- {
		if s[i] == '(' {
			left++
		} else {
			right++
		}
		if left == right {
			maxLen = max(maxLen, 2*left)
		} else if right < left {
			left = 0
			right = 0
		}
	}
	return maxLen
}
